explain: add indexes (#3933)
pingcap/docs: explain-indexes.md

3 rows in set (0.00 sec)

+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
| id                       | estRows | task      | access object                  | operator info                     |
+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
| Projection_4             | 1.00    | root      |                                | test.t1.id                        |
| └─IndexReader_6          | 1.00    | root      |                                | index:IndexRangeScan_5            |
|   └─IndexRangeScan_5     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | range:[123,123], keep order:false |
+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
3 rows in set (0.00 sec)
```

Because `id` is also the internal `RowID`, it is stored in the `intkey` index. After using the `intkey` index as part of `└─IndexRangeScan_5`, the value of the `RowID` can be returned directly.

## Point_Get and Batch_Point_Get

> The `Point_Get` and `Batch_Point_Get` operator is used when retrieving data directly from the Primary Key. This operator is differs from `IndexLookup` in that all data can be retrieved in a single-step: 

Reviewer: Not only the primary key. A unique key could satisfy the `Point_Get`.

Author: I think this is only the case where the first unique not null key is promoted to the primary key. Otherwise, the unique table row will be on a separate region to the index. I verified this with the following testcase:
```sql
mysql> show create table t1\G
*************************** 1. row ***************************
       Table: t1
Create Table: CREATE TABLE `t1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `intkey` int(11) NOT NULL,
  `pad1` varbinary(1024) DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `intkey` (`intkey`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=2000001
1 row in set (0.00 sec)

mysql> ALTER TABLE t1 add unique_key int;
Query OK, 0 rows affected (0.26 sec)

mysql> update t1 set unique_key = id;
Query OK, 1010 rows affected (0.05 sec)
Rows matched: 1010  Changed: 1010  Warnings: 0

mysql> alter table t1 add index (unique_key);
Query OK, 0 rows affected (0.33 sec)

mysql> explain select * from t1 where unique_key = 1234;
+-------------------------------+---------+-----------+----------------------------------------+-------------------------------------+
| id                            | estRows | task      | access object                          | operator info                       |
+-------------------------------+---------+-----------+----------------------------------------+-------------------------------------+
| IndexLookUp_10                | 0.00    | root      |                                        |                                     |
| ├─IndexRangeScan_8(Build)     | 0.00    | cop[tikv] | table:t1, index:unique_key(unique_key) | range:[1234,1234], keep order:false |
| └─TableRowIDScan_9(Probe)     | 0.00    | cop[tikv] | table:t1                               | keep order:false                    |
+-------------------------------+---------+-----------+----------------------------------------+-------------------------------------+
3 rows in set (0.00 sec)
```

@lzmhhh123 can you confirm what you meant if otherwise?

Reviewer 2: @nullnotnil That's right.

Reviewer 2: I think it's buggy after altering table.

```mysql
mysql> CREATE TABLE `t1` (
    ->   `id` int(11) NOT NULL AUTO_INCREMENT,
    ->   `intkey` int(11) NOT NULL,
    ->   `pad1` varbinary(1024) DEFAULT NULL,
    ->   unique_key int,
    ->   PRIMARY KEY (`id`),
    ->   KEY `intkey` (`intkey`),
    ->   UNIQUE KEY (unique_key)
    -> )
    -> ;
Query OK, 0 rows affected (0.00 sec)

mysql> explain select * from t1 where unique_key = 1234;
+-------------+---------+------+----------------------------------------+---------------+
| id          | estRows | task | access object                          | operator info |
+-------------+---------+------+----------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:t1, index:unique_key(unique_key) |               |
+-------------+---------+------+----------------------------------------+---------------+
1 row in set (0.00 sec)
```

Reviewer 2: And here's the condition to use Point Get, https://github.com/pingcap/tidb/blob/adb3a4a42f05e02afbe2667771229640a6434851/planner/core/point_get_plan.go#L645-L651

Author: nvm, there is a bug in my testcase :-) I created a non-unique key. I can reproduce:

```sql
mysql> DROP TABLE IF EXISTS t1;
Query OK, 0 rows affected (0.01 sec)

mysql> CREATE TABLE `t1` (
    ->   `id` int(11) NOT NULL AUTO_INCREMENT,
    ->   `intkey` int(11) NOT NULL,
    ->   `pad1` varbinary(1024) DEFAULT NULL,
    ->   PRIMARY KEY (`id`),
    ->   KEY `intkey` (`intkey`)
    -> ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=2000001;
Query OK, 0 rows affected (0.01 sec)

mysql> 
mysql> ALTER TABLE t1 ADD unique_key INT;
Query OK, 0 rows affected (0.01 sec)

mysql> UPDATE t1 SET unique_key = id;
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0  Changed: 0  Warnings: 0

mysql> ALTER TABLE t1 ADD UNIQUE KEY (unique_key);
Query OK, 0 rows affected (0.01 sec)

mysql> 
mysql> explain select * from t1 where unique_key = 1234;
+-------------+---------+------+----------------------------------------+---------------+
| id          | estRows | task | access object                          | operator info |
+-------------+---------+------+----------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:t1, index:unique_key(unique_key) |               |
+-------------+---------+------+----------------------------------------+---------------+
1 row in set (0.00 sec)
```

Author: Thanks! I have improved in this commit: https://github.com/pingcap/docs/pull/3933/commits/39ee9ce055e00c468b29c9a067d11748d61fe978

5 rows in set (0.00 sec)

+------------------------------+---------+-----------+--------------------------------+-------------------------------------+
| id                           | estRows | task      | access object                  | operator info                       |
+------------------------------+---------+-----------+--------------------------------+-------------------------------------+
| StreamAgg_12                 | 1.00    | root      |                                | funcs:max(test.t1.intkey)->Column#4 |
| └─Limit_16                   | 1.00    | root      |                                | offset:0, count:1                   |
|   └─IndexReader_29           | 1.00    | root      |                                | index:Limit_28                      |
|     └─Limit_28               | 1.00    | cop[tikv] |                                | offset:0, count:1                   |
|       └─IndexFullScan_27     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | keep order:true, desc               |
+------------------------------+---------+-----------+--------------------------------+-------------------------------------+
5 rows in set (0.00 sec)
```

> In the above statements, an `IndexFullScan` is performed on each TiKV region. Despite the name `FullScan`, only the first row needs to be read (`└─Limit_28`). Each TiKV region returns its `MIN` or `MAX` value to TiDB, which then performs Stream Aggregation to filter for a single row.

Reviewer 2: Stream Aggregation with the aggregation function `MAX` or `MIN` also ensures returning `NULL` if the table is empty.

Author: https://github.com/pingcap/docs/pull/3933/commits/0640418ec72b189d3651c340cedf23d5b01ff302

> By contrast, executing `MIN` on an unindexed value will result in a `TableFullScan`. The query requires all rows to be scanned in TiKV, but a `TopN` is performed to ensure each TiKV region only returns one row to TiDB:

Reviewer 2: Should we tell the user the performance difference that `TopN` must scan the whole table and is much less effective than Limit with Index?

Author: Added to https://github.com/pingcap/docs/pull/3933/commits/0640418ec72b189d3651c340cedf23d5b01ff302

+----------------------------+---------+-----------+--------------------------------+----------------------------------------------------------------------------+
| StreamAgg_20               | 1.00    | root      |                                | funcs:avg(Column#7, Column#8)->Column#4                                    |
| └─IndexReader_21           | 1.00    | root      |                                | index:StreamAgg_8                                                          |
|   └─StreamAgg_8            | 1.00    | cop[tikv] |                                | funcs:count(test.t1.intkey)->Column#7, funcs:sum(test.t1.intkey)->Column#8 |
|     └─IndexFullScan_19     | 1010.00 | cop[tikv] | table:t1, index:intkey(intkey) | keep order:false                                                           |
+----------------------------+---------+-----------+--------------------------------+----------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

This is considered more efficient than a `TableFullScan` because the width of the value in the index `(intkey + RowID)` is less than the width of the full row.

> The following statement does not support using an `IndexFullScan` because additional rows are required from the table:

Reviewer 2: ```suggestion
The following statement does not support using an `IndexFullScan` because additional columns are required from the table:
```

---
title: Explain Statements That Use Indexes
summary: Learn about the execution plan information returned by the EXPLAIN statement in TiDB.
---

# Explain Statements That Use Indexes

TiDB supports several operators which make use of indexes to speed up query execution. The examples in this document are based on the following sample data:

{{< copyable "sql" >}}

```sql
CREATE TABLE t1 (
 id INT NOT NULL PRIMARY KEY auto_increment,
 intkey INT NOT NULL,
 pad1 VARBINARY(1024),
 INDEX (intkey)
);

INSERT INTO t1 SELECT NULL, FLOOR(RAND()*1024), RANDOM_BYTES(1024) FROM dual;
INSERT INTO t1 SELECT NULL, FLOOR(RAND()*1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
INSERT INTO t1 SELECT NULL, FLOOR(RAND()*1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
INSERT INTO t1 SELECT NULL, FLOOR(RAND()*1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
```

## IndexLookup operator

> The `IndexLookup` operator is used by TiDB when retrieving data from a secondary index. In this case, the following queries will all use the `IndexLookup` operator on the index `intkey`:

Reviewer 3: ```suggestion
TiDB uses the `IndexLookup` operator when retrieving data from a secondary index. In this case, the following queries will all use the `IndexLookup` operator on the `intkey` index:
```

+-------------------------------+---------+-----------+--------------------------------+-----------------------------------------------------+
3 rows in set (0.00 sec)

+-------------------------------+---------+-----------+--------------------------------+----------------------------------+
| id                            | estRows | task      | access object                  | operator info                    |
+-------------------------------+---------+-----------+--------------------------------+----------------------------------+
| IndexLookUp_10                | 6.00    | root      |                                |                                  |
| ├─IndexRangeScan_8(Build)     | 6.00    | cop[tikv] | table:t1, index:intkey(intkey) | range:[99,103], keep order:false |
| └─TableRowIDScan_9(Probe)     | 6.00    | cop[tikv] | table:t1                       | keep order:false                 |
+-------------------------------+---------+-----------+--------------------------------+----------------------------------+
3 rows in set (0.00 sec)
```

The `IndexLookup` operator has two child nodes:

> * The `├─IndexRangeScan_8(Build)` performs a range scan on the `intkey` index and retries the value of the internal `RowID`, which for this table is also the `PRIMARY KEY`.

Reviewer 3: ```suggestion
* The `├─IndexRangeScan_8(Build)` operator performs a range scan on the `intkey` index and retrieves the values of the internal `RowID` (for this table, the primary key).
```

* The `└─TableRowIDScan_9(Probe)` operator then retrieves the full row from the table data.

> Because an `IndexLookup` requires two steps, the SQL Optimizer may elect to use the `TableFullScan` operator in scenarios where a large number of rows match. The optimizer uses [statistics](/statistics.md) in order to make these decisions. In the following example, a large number of rows match the condition of `intkey > 100`, and a `TableFullScan` is chosen:

Reviewer 3: ```suggestion
Because an `IndexLookup` task requires two steps, the SQL Optimizer might choose the `TableFullScan` operator based on [statistics](/statistics.md) in scenarios where a large number of rows match. In the following example, a large number of rows match the condition of `intkey > 100`, and a `TableFullScan` is chosen:
```

{{< copyable "sql" >}}

```sql
EXPLAIN SELECT * FROM t1 ORDER BY intkey DESC LIMIT 10;
```

```sql
+--------------------------------+---------+-----------+--------------------------------+------------------------------------+
| id                             | estRows | task      | access object                  | operator info                      |
+--------------------------------+---------+-----------+--------------------------------+------------------------------------+
| IndexLookUp_21                 | 10.00   | root      |                                | limit embedded(offset:0, count:10) |
| ├─Limit_20(Build)              | 10.00   | cop[tikv] |                                | offset:0, count:10                 |
| │ └─IndexFullScan_18           | 10.00   | cop[tikv] | table:t1, index:intkey(intkey) | keep order:true, desc              |
| └─TableRowIDScan_19(Probe)     | 10.00   | cop[tikv] | table:t1                       | keep order:false, stats:pseudo     |
+--------------------------------+---------+-----------+--------------------------------+------------------------------------+
4 rows in set (0.00 sec)

```

> In the above example, the last 20 rows are read from the index `intkey`. These `RowID` values are then retrieved from the table data.

Reviewer 3: ```suggestion
In the above example, the last 20 rows are read from the `intkey` index. These `RowID` values are then retrieved from the table data.
```

## IndexReader

> TiDB supports the _covering index optimization_. If all rows can be retrieved from an index, TiDB will skip the second step that is usually required in an `IndexLookup`. Consider the following two examples:

Reviewer 3: ```suggestion
TiDB supports the _covering index optimization_. If all rows can be retrieved from an index, TiDB will skip the second step that is usually required in an `IndexLookup` task. Consider the following two examples:
```

| └─TableRowIDScan_9(Probe)     | 1.00    | cop[tikv] | table:t1                       | keep order:false                  |
+-------------------------------+---------+-----------+--------------------------------+-----------------------------------+
3 rows in set (0.00 sec)

+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
| id                       | estRows | task      | access object                  | operator info                     |
+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
| Projection_4             | 1.00    | root      |                                | test.t1.id                        |
| └─IndexReader_6          | 1.00    | root      |                                | index:IndexRangeScan_5            |
|   └─IndexRangeScan_5     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | range:[123,123], keep order:false |
+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
3 rows in set (0.00 sec)
```

> Because `id` is also the internal `RowID`, it is stored in the `intkey` index. After using the `intkey` index as part of `└─IndexRangeScan_5`, the value of the `RowID` can be returned directly.

Reviewer 3: ```suggestion
Because `id` is also the internal `RowID`, `id` is stored in the `intkey` index. After using the `intkey` index as a part of the `└─IndexRangeScan_5` operator task, the value of the `RowID` can be returned directly.
```

## Point_Get and Batch_Point_Get

> The `Point_Get` and `Batch_Point_Get` operator is used when retrieving data directly from a Primary Key or Unique Key. This operator is more efficient than `IndexLookup`. For example:

Reviewer 3: ```suggestion
TiDB uses the `Point_Get` or `Batch_Point_Get` operator when retrieving data directly from a primary key or unique key. The two operators are more efficient than `IndexLookup`. For example:
```

> The `Point_Get` and `Batch_Point_Get` operator is used when retrieving data directly from a Primary Key or Unique Key. This operator is more efficient than `IndexLookup`, but only applies to `SELECT` statements against a single table. For example:

Reviewer: ```suggestion
The `Point_Get` and `Batch_Point_Get` operator is used when retrieving data directly from a Primary Key or Unique Key. This operator is more efficient than `IndexLookup`. For example:
```

Reviewer: ```sql
tidb(localhost:4000) > drop table if exists t1, t2;
Query OK, 0 rows affected (0.02 sec)

tidb(localhost:4000) > create table t1 (a int, b int, primary key(a));
Query OK, 0 rows affected (0.00 sec)

tidb(localhost:4000) > create table t2 (a varchar(20), unique key(a));
Query OK, 0 rows affected (0.01 sec)

tidb(localhost:4000) > explain select * from t1 join t2 where t1.a = 1 and t2.a = 1 and t2.a = "something";
+-----------------------------+---------+------+----------------------+---------------------------------+
| id                          | estRows | task | access object        | operator info                   |
+-----------------------------+---------+------+----------------------+---------------------------------+
| Projection_7                | 0.80    | root |                      | test.t1.a, test.t1.b, test.t2.a |
| └─HashJoin_9                | 0.80    | root |                      | CARTESIAN inner join            |
|   ├─Selection_11(Build)     | 0.80    | root |                      | eq(cast(test.t2.a), 1)          |
|   │ └─Point_Get_10          | 1.00    | root | table:t2, index:a(a) |                                 |
|   └─Point_Get_12(Probe)     | 1.00    | root | table:t1             | handle:1                        |
+-----------------------------+---------+------+----------------------+---------------------------------+
5 rows in set (0.00 sec)
```

Rows matched: 1010  Changed: 1010  Warnings: 0

Query OK, 0 rows affected (0.37 sec)

+-------------+---------+------+----------------------------------------+---------------+
| id          | estRows | task | access object                          | operator info |
+-------------+---------+------+----------------------------------------+---------------+
| Point_Get_1 | 1.00    | root | table:t1, index:unique_key(unique_key) |               |
+-------------+---------+------+----------------------------------------+---------------+
1 row in set (0.00 sec)

+-------------------+---------+------+----------------------------------------+------------------------------+
| id                | estRows | task | access object                          | operator info                |
+-------------------+---------+------+----------------------------------------+------------------------------+
| Batch_Point_Get_1 | 2.00    | root | table:t1, index:unique_key(unique_key) | keep order:false, desc:false |
+-------------------+---------+------+----------------------------------------+------------------------------+
1 row in set (0.00 sec)
```

## IndexFullScan

> Because indexes are ordered, the `IndexFullScan` can be used to optimize common queries such as the `MIN` or `MAX` values for an indexed value:

Reviewer 3: ```suggestion
Because indexes are ordered, the `IndexFullScan` operator can be used to optimize common queries such as the `MIN` or `MAX` values for an indexed value:
```

5 rows in set (0.00 sec)

+------------------------------+---------+-----------+--------------------------------+-------------------------------------+
| id                           | estRows | task      | access object                  | operator info                       |
+------------------------------+---------+-----------+--------------------------------+-------------------------------------+
| StreamAgg_12                 | 1.00    | root      |                                | funcs:max(test.t1.intkey)->Column#4 |
| └─Limit_16                   | 1.00    | root      |                                | offset:0, count:1                   |
|   └─IndexReader_29           | 1.00    | root      |                                | index:Limit_28                      |
|     └─Limit_28               | 1.00    | cop[tikv] |                                | offset:0, count:1                   |
|       └─IndexFullScan_27     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | keep order:true, desc               |
+------------------------------+---------+-----------+--------------------------------+-------------------------------------+
5 rows in set (0.00 sec)
```

> In the above statements, an `IndexFullScan` is performed on each TiKV region. Despite the name `FullScan`, only the first row needs to be read (`└─Limit_28`). Each TiKV region returns its `MIN` or `MAX` value to TiDB, which then performs Stream Aggregation to filter for a single row. Stream Aggregation with the aggregation function `MAX` or `MIN` also ensures returning `NULL` if the table is empty.

Reviewer 3: ```suggestion
In the above statements, an `IndexFullScan` task is performed on each TiKV Region. Despite the name `FullScan`, only the first row needs to be read (`└─Limit_28`). Each TiKV Region returns its `MIN` or `MAX` value to TiDB, which then performs Stream Aggregation to filter for a single row. Stream Aggregation with the aggregation function `MAX` or `MIN` also ensures that `NULL` is returned if the table is empty.
```

> By contrast, executing `MIN` on an unindexed value will result in a `TableFullScan`. The query requires all rows to be scanned in TiKV, but a `TopN` is performed to ensure each TiKV region only returns one row to TiDB. While the `TopN` prevents excessive rows being transfered between TiKV and TiDB, this statement is still considered far less efficient than the above example, where `MIN` is able to make use of an index.

Reviewer 3: ```suggestion
By contrast, executing the `MIN` function on an unindexed value will result in `TableFullScan`. The query will require all rows to be scanned in TiKV, but a `TopN` calculation is performed to ensure each TiKV Region only returns one row to TiDB. Although `TopN` prevents excessive rows from being transferred between TiKV and TiDB, this statement is still considered far less efficient than the above example where `MIN` is able to make use of an index.
```

| id                         | estRows | task      | access object                  | operator info                                                              |
+----------------------------+---------+-----------+--------------------------------+----------------------------------------------------------------------------+
| StreamAgg_20               | 1.00    | root      |                                | funcs:avg(Column#7, Column#8)->Column#4                                    |
| └─IndexReader_21           | 1.00    | root      |                                | index:StreamAgg_8                                                          |
|   └─StreamAgg_8            | 1.00    | cop[tikv] |                                | funcs:count(test.t1.intkey)->Column#7, funcs:sum(test.t1.intkey)->Column#8 |
|     └─IndexFullScan_19     | 1010.00 | cop[tikv] | table:t1, index:intkey(intkey) | keep order:false                                                           |
+----------------------------+---------+-----------+--------------------------------+----------------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

> This is considered more efficient than a `TableFullScan` because the width of the value in the index `(intkey + RowID)` is less than the width of the full row.

Reviewer 3: ```suggestion
In the above examples, `IndexFullScan` is more efficient than `TableFullScan` because the width of the value in the `(intkey + RowID)` index is less than the width of the full row.
```

> The following statement does not support using an `IndexFullScan` because additional columns are required from the table:

Reviewer 3: ```suggestion
The following statement does not support using an `IndexFullScan` operator because additional columns are required from the table:
```

3 rows in set (0.00 sec)

+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
| id                       | estRows | task      | access object                  | operator info                     |
+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
| Projection_4             | 1.00    | root      |                                | test.t1.id                        |
| └─IndexReader_6          | 1.00    | root      |                                | index:IndexRangeScan_5            |
|   └─IndexRangeScan_5     | 1.00    | cop[tikv] | table:t1, index:intkey(intkey) | range:[123,123], keep order:false |
+--------------------------+---------+-----------+--------------------------------+-----------------------------------+
3 rows in set (0.00 sec)
```

Because `id` is also the internal `RowID`, it is stored in the `intkey` index. After using the `intkey` index as part of `└─IndexRangeScan_5`, the value of the `RowID` can be returned directly.

## Point_Get and Batch_Point_Get

> The `Point_Get` and `Batch_Point_Get` operator is used when retrieving data directly from a Primary Key or Unique Key. This operator is more efficient than `IndexLookup`. For example:

Reviewer 3: ```suggestion
TiDB uses the `Point_Get` or `Batch_Point_Get` operator when retrieving data directly from a primary key or unique key. These operators are more efficient than `IndexLookup`. For example:
```
